domain/controller/service/service.go

```diff
@@ -6,27 +6,87 @@ package service
 import (
 	"context"
 
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/database"
 	"github.com/juju/juju/core/model"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+	"github.com/juju/juju/internal/logger"
 )
 
+var log = logger.GetLogger("juju.domain.controller.service")
+
 // State defines an interface for interacting with the underlying state.
 type State interface {
 	ControllerModelUUID(ctx context.Context) (model.UUID, error)
+	GetModelActivationStatus(ctx context.Context, controllerModelUUID string) (bool, error)
+	AllModelActivationStatusQuery() string
+}
+
+// WatcherFactory describes methods for creating watchers.
+type WatcherFactory interface {
+	// NewNamespaceWatcher returns a new namespace watcher
+	// for events based on the input change mask.
+	NewNamespaceMapperWatcher(
+		namespace string, changeMask changestream.ChangeType, initialStateQuery eventsource.NamespaceQuery, mapper eventsource.Mapper,
+	) (watcher.StringsWatcher, error)
+}
+
+type WatcherFactoryGetter interface {
+	GetWatcherFactory() WatcherFactory
 }
 
 // Service defines a service for interacting with the underlying state.
 type Service struct {
-	st State
+	st             State
+	watcherFactory WatcherFactory
 }
 
 // NewService returns a new Service for interacting with the underlying state.
-func NewService(st State) *Service {
+func NewService(st State, watcherFactory WatcherFactory) *Service {
 	return &Service{
-		st: st,
+		st:             st,
+		watcherFactory: watcherFactory,
 	}
 }
 
 // ControllerModelUUID returns the model UUID of the controller model.
 func (s *Service) ControllerModelUUID(ctx context.Context) (model.UUID, error) {
 	return s.st.ControllerModelUUID(ctx)
 }
+
+// Watch returns a watcher that monitors changes to models.
+func (s *Service) Watch(ctx context.Context) (watcher.StringsWatcher, error) {
+	mapper := func(ctx context.Context, db database.TxnRunner, changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
```

Comment:
I have had a look at the model worker manager and what its requirements are. We don't need to concern ourselves with the model delete events here as that is handled separately.

The logic for this really just needs to check the model uuid's are active and if so pass along the change.

I suspect we also might want to look at making the state call that we use here a bulk state call. i.e consider if it would be possible to pass in a set of model uuid's and just receive back the subset of those uuid's that are still active.

We don't care if a model is deleted, not activated or not found in here. We can just throw the event away in this case.

Reply:
Hi Tom, might have misunderstood on the delete part in our previous convo. Sure I could look at bulk state call, I agree it would be more efficient to filter the active status after retrieving from state